### PR TITLE
ceph: add expansion support for encrypted osd on pvc

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

* ceph: support osd deployment update on encrypted pvc
If the deployment spec was updated or restarted, the "encryption-open"
init container would fail to open the device since it was already
opened on the node.
Because they are no reliable post stop hook on Kubernetes, the trick is
to use a small bash init container.

* ceph: add pvc expansion support for encrypted osds
We now support the expansion of encrypted osds. We added a new init
container to run "cryptsetup resize".

* ceph: add encrypted block status container
This init container displays the status of the encrypted pvc. This can
be really useful for debugging purpose.

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by leseb** [comment]
Tested on OCP, previously the PVC was 10GiB and was updated to 20GiB:

```
[root@rook-ceph-osd-1-7c94db5df8-82tpc /]# lsblk 
NAME                                                            MAJ:MIN   RM  SIZE RO TYPE  MOUNTPOINT
xvdbw                                                           202:18944  0   10G  0 disk  
`-ceph-e22248a7-9397-4828-87c0-1213e0909e41-xvdbw-block-dmcrypt 253:0      0   10G  0 crypt 
xvda                                                            202:0      0  120G  0 disk  
|-xvda2                                                         202:2      0    1G  0 part  
|-xvda3                                                         202:3      0  119G  0 part  /var/lib/ceph/crash
`-xvda1                                                         202:1      0    1M  0 part  
loop0                                                             7:0      0   20G  0 loop  
xvdcl                                                           202:22784  0   20G  0 disk  
`-set1-data-0-7swr7-block-dmcrypt                               253:1      0   20G  0 crypt 
xvdcf                                                           202:21248  0    8G  0 disk  
[root@rook-ceph-osd-1-7c94db5df8-82tpc /]# exit
[leseb@tarox~/go/src/github.com/rook/rook][fix-5996 !] 
[leseb@tarox~/go/src/github.com/rook/rook][fix-5996 !] 
[leseb@tarox~/go/src/github.com/rook/rook][fix-5996 !] oc logs rook-ceph-osd-1-7c94db5df8-82tpc -c expand-encrypted-bluefs
Command successful.
[leseb@tarox~/go/src/github.com/rook/rook][fix-5996 !] oc logs rook-ceph-osd-1-7c94db5df8-82tpc -c expand-bluefs
inferring bluefs devices from bluestore path
1 : device size 0x4ffe00000 : own 0x[11ff00000~40000000] = 0x40000000 : using 0x470000(4.4 MiB) : bluestore has 0x23fdd0000(9.0 GiB) available
Expanding DB/WAL...
Expanding Main...
1 : expanding  from 0x27fe00000 to 0x4ffe00000
1 : size label updated to 21472739328
```

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 1]
```diff
@@ -801,6 +826,53 @@ func (c *Cluster) getExpandPVCInitContainer(osdProps osdProperties, osdID string
...
+			"cryptsetup",
+		},
+		Args:            []string{"--verbose", "resize", encryptionDMName(osdProps.pvc.ClaimName, DmcryptBlockType)},
+		VolumeMounts:    []v1.VolumeMount{getPvcOSDBridgeMountActivate(mountPath, osdProps.pvc.ClaimName)},
+		SecurityContext: PrivilegedContext(),
+		Resources:       osdProps.resources,
+	}
+}
+
+func (c *Cluster) getEncryptedStatusPVCInitContainer(mountPath string, osdProps osdProperties) v1.Container {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go:850]: Why do we need this init container? Is it just to log the status? If it's just for debugging, doesn't seem like we should have it in production.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 2]
```diff
@@ -332,9 +350,14 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 
 	if osdProps.onPVC() && osd.CVMode == "raw" {
 		if osdProps.encrypted {
-			// Add a new init container to open the encrypted disk!
+			// Open the encrypted disk
 			initContainers = append(initContainers, c.getPVCEncryptionOpenInitContainerActivate(osdProps)...)
+			// Copy the encrypted block to the osd data location, e,g: /var/lib/ceph/osd/ceph-0/block
 			initContainers = append(initContainers, c.getPVCEncryptionInitContainerActivate(osdDataDirPath, osdProps)...)
+			// Print the encrypted block status
+			initContainers = append(initContainers, c.getEncryptedStatusPVCInitContainer(osdDataDirPath, osdProps))
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go:358]: Is there a unit test that covers creating the encrypted pod spec?

Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go:358]: Not really. I can take care of that.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go:850]: I think this is useful to validate a few things, such as the encryption expansion has worked correctly. I think it's a nice to have, not necessarily a must-have though.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 2 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go:358]: Actually, I've added a few during the initial implementation, I can try to add more.

Comment (leseb) [reply pkg/operator/ceph/cluster/osd/spec.go:358]: Added!

**Pull request merged**